DepositRegistrar.sol

```diff
@@ -0,0 +1,306 @@
+        uint n = averagePeriod < (60 seconds * M) ? 7 days * M / averagePeriod : 10000;
+        averagePrice = ((n-1)*averagePrice + name.value ) / n;
+        name.averagePrice = averagePrice;
+        lastSinceNewName = now;
+        
+        Deed deedContract = Deed(name.deed);
+        
+        deedContract.setBalance(name.value > minBid ? name.value : minBid);
+    }
+    
+    
+    /*
+    ## Renewals can be done at any moment by renewing the deposit
+
+    In order to renew a name you need to calculate how much have average prices of 
+    names changed since you last bought it. If the average price names have increased, 
+    then you need to increase your deposit. If median prices have fallen down, then 
+    you have the right to withdraw part of the difference (1/2 after a year, 3/4 
+    after 2 years, 7/8 after 3 years etc). 
+    */ 
+    function updatedValue(bytes32 nameHash) constant returns (uint updatedPrice) {
+        entry name = hashedNames[nameHash];
+        return name.value * averagePrice / name.averagePrice;
+    }
+    
+    function renewDeed(bytes32 nameHash) {
+        entry name = hashedNames[nameHash];
+        if (now < name.firstRegistered + renewalPeriod/2 ) throw;
+        Deed deedContract = Deed(name.deed);
+        uint difference = 0;
+        if (msg.sender != deedContract.owner() && name.status != Mode.Owned) throw;
+
+        uint updatedPrice = updatedValue(nameHash);
+        
+        if (updatedPrice > name.value) {
+            difference = updatedPrice - name.value; 
+            if (msg.value < difference) throw;
+            deedContract.setBalance.value(msg.value)(updatedPrice);
+        } else {
+            difference =  name.value - updatedPrice; 
+            uint ratioOfRecovery = 2**((now - name.lastRenewed)/ 1 years);
+            // the more time, the more you recover more from your initial investment
+            deedContract.setBalance(name.value + difference - difference/ratioOfRecovery);
+        }
+        
+        name.value = updatedPrice;
+        name.lastRenewed = now;
+        // next renewal data is twice the current age
+        name.renewalDate = 2 * now - name.firstRegistered;
+    }
+     
+    /*
+    ## Renewals can be done at any moment by renewing the deposit
+
+    The real cost of holding a name is the opportunity cost of doing something better with 
+    your ether. If there are better opportunities, like staking it, lending it or investing 
+    in some other new venture, then holding names should be seen as an undesirable outcome 
+    and owners have an incentive to release them. Names can be sold at any moment, but the 
+    buyer will incur the same renewal cost/benefit analysis. 
+    */    
+    function releaseDeed(bytes32 nameHash) noEther  {
+        entry name = hashedNames[nameHash];
+        Deed deedContract = Deed(name.deed);
+        if (now < name.firstRegistered + renewalPeriod ) throw;
```

Comment:
This is where the restriction should be added to half the renewalPeriod. Otherwise someone could register dummy names at almost no cost to manipulate the averages.